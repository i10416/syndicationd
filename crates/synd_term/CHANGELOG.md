# Changelog

All notable changes to this project will be documented in this file.

## [unreleased]

### Features

- Add baggage propagation by [@ymgyt](https://github.com/ymgyt) ([d02e514c](https://github.com/ymgyt/syndicationd/commit/d02e514c8f6e32aa748c10dadb204153cba21ecc))
- Add opentelemetry layers by [@ymgyt](https://github.com/ymgyt) ([4d3f5bf3](https://github.com/ymgyt/syndicationd/commit/4d3f5bf3f45f31cfd014dbdf37a41a31ea0472ca))
- Update ratatui to 0.26 and fix breaking apis by [@ymgyt](https://github.com/ymgyt) ([c482683a](https://github.com/ymgyt/syndicationd/commit/c482683a0083baf93a60ef31b280c49ac4eafccb))
- Change local time format by [@ymgyt](https://github.com/ymgyt) ([fb826165](https://github.com/ymgyt/syndicationd/commit/fb826165367eb97c0bec216db286bf1ee13fba07))
- Use tailwind color palettes by [@ymgyt](https://github.com/ymgyt) ([a93b8ec7](https://github.com/ymgyt/syndicationd/commit/a93b8ec753d3f0da2c4915cc258b3b1054ccef57))
- Change entries table constraint by [@ymgyt](https://github.com/ymgyt) ([53027a59](https://github.com/ymgyt/syndicationd/commit/53027a59aa1bb8c24deeb5696dac52f2704104bc))
- Add palette flag by [@ymgyt](https://github.com/ymgyt) ([04dc486d](https://github.com/ymgyt/syndicationd/commit/04dc486d0ab3043e021e164e70f5fe081e3c464d))
- Add in_flight by [@ymgyt](https://github.com/ymgyt) ([eae48336](https://github.com/ymgyt/syndicationd/commit/eae48336cc6e5298bc6c78599fa3054a134a170e))
- Add in flight throbber by [@ymgyt](https://github.com/ymgyt) ([fef77519](https://github.com/ymgyt/syndicationd/commit/fef77519e2ca59e5d267d6cecab8c008e92adc2c))
- Add instrument by [@ymgyt](https://github.com/ymgyt) ([dfbe9350](https://github.com/ymgyt/syndicationd/commit/dfbe93501542ff75361ddf3b158e21f7e77329b3))
- Impl kvsd client by [@ymgyt](https://github.com/ymgyt) ([6ae6de7a](https://github.com/ymgyt/syndicationd/commit/6ae6de7a2e783417b1a8d5d3c2b450109d83725f))
- Improve subscription input handling by [@ymgyt](https://github.com/ymgyt) ([309d8fac](https://github.com/ymgyt/syndicationd/commit/309d8fac0ea33438af61df374f32a73e235ec63f))
- Improve feed subscription flow by [@ymgyt](https://github.com/ymgyt) ([088d18df](https://github.com/ymgyt/syndicationd/commit/088d18df15486d4635a5dc2014f62b9fce6a9db6))
- Swap terminal restore step by [@ymgyt](https://github.com/ymgyt) ([2f9f2cb7](https://github.com/ymgyt/syndicationd/commit/2f9f2cb7830d7cb473b847f1969c9125428e4a6e))
- Remove unsubscribed entries by [@ymgyt](https://github.com/ymgyt) ([d29ba92e](https://github.com/ymgyt/syndicationd/commit/d29ba92e929d9d1348fa114ac2bdf210b76c5a1b))
- Reload entries when subscribe feed by [@ymgyt](https://github.com/ymgyt) ([6e0aa72b](https://github.com/ymgyt/syndicationd/commit/6e0aa72b67a17e7139b532940c24f70a7642a39d))
- Serve https by [@ymgyt](https://github.com/ymgyt) ([fbb9011e](https://github.com/ymgyt/syndicationd/commit/fbb9011e86acf6e4cf30f37a74e67d3202bbc5a0))
- Support axum_server graceful shutdown by [@ymgyt](https://github.com/ymgyt) ([880b6d3e](https://github.com/ymgyt/syndicationd/commit/880b6d3e8d0f90b711a1d6e8e1bf6fb1808e5161))
- Use cow by [@ymgyt](https://github.com/ymgyt) ([ab6ae298](https://github.com/ymgyt/syndicationd/commit/ab6ae298abeda1d7d3c67939bc70f0d2269e8654))
- Update default endpoint by [@ymgyt](https://github.com/ymgyt) ([e684b0cc](https://github.com/ymgyt/syndicationd/commit/e684b0cc4122a3fd4ece6a1e3697f71aaa311daf))

### Bug Fixes

- Workarround scrollbar rendering bug by [@ymgyt](https://github.com/ymgyt) ([d2982cb6](https://github.com/ymgyt/syndicationd/commit/d2982cb6c8fa385655290d953aa9243d3470382d))
- Build by [@ymgyt](https://github.com/ymgyt) ([bd340e9d](https://github.com/ymgyt/syndicationd/commit/bd340e9d30f101c891f53b2d2be10a0cf8833f4b))

### Miscellaneous Tasks

- Format toml by [@ymgyt](https://github.com/ymgyt) ([36677745](https://github.com/ymgyt/syndicationd/commit/3667774506106fe0f38d77efac9f4b27c70090aa))
- Update CHANGELOG by [@ymgyt](https://github.com/ymgyt) ([f5091f3c](https://github.com/ymgyt/syndicationd/commit/f5091f3ceff04b9ff818bb4e0ce0e4bbe9851177))
- Update CHANGELOG by [@ymgyt](https://github.com/ymgyt) ([99f036df](https://github.com/ymgyt/syndicationd/commit/99f036dfe227c1670f967aa949116e3ae8a2c97b))
- Use hyphen as package name instead of underscore by [@ymgyt](https://github.com/ymgyt) ([0a8ed059](https://github.com/ymgyt/syndicationd/commit/0a8ed05997790f9f05c932c92fa2b2b2d74065a9))
- Update CHANGELOG by [@ymgyt](https://github.com/ymgyt) ([e688670c](https://github.com/ymgyt/syndicationd/commit/e688670c853718a1cb825cb787861dffe55046d1))
- Rename synd-authn to synt-auth to publish as a new crate by [@ymgyt](https://github.com/ymgyt) ([59ae4ffa](https://github.com/ymgyt/syndicationd/commit/59ae4ffa51f5323fa4a3aae5e30e950b15730519))
- Update CHANGELOG by [@ymgyt](https://github.com/ymgyt) ([31eb8a34](https://github.com/ymgyt/syndicationd/commit/31eb8a3472e770931fab427e2a8c74a9754b157a))
- Update CHANGELOG by [@ymgyt](https://github.com/ymgyt) ([e1910ce1](https://github.com/ymgyt/syndicationd/commit/e1910ce120ca9d8f38fa4c479156f723d54ae59c))

### Refactor

- Rename crates by [@ymgyt](https://github.com/ymgyt) ([ce0982e4](https://github.com/ymgyt/syndicationd/commit/ce0982e497647b23dcf07e39d525121bcd9ac1fa))
- Create synd_authn crate by [@ymgyt](https://github.com/ymgyt) ([682bcc6f](https://github.com/ymgyt/syndicationd/commit/682bcc6ff3c035be566dea99d2487e0173537c8d))
- Use clippy pedantic by [@ymgyt](https://github.com/ymgyt) ([328ddade](https://github.com/ymgyt/syndicationd/commit/328ddadebbad5381271c5e84cce2d6888252e70c))
- Clippy by [@ymgyt](https://github.com/ymgyt) ([a1693b36](https://github.com/ymgyt/syndicationd/commit/a1693b36b73ad3987af9a853e214392d8b1eae8d))
- Fix lint by [@ymgyt](https://github.com/ymgyt) ([aac00b98](https://github.com/ymgyt/syndicationd/commit/aac00b98335bb75cc57fdea0875bfd675bf8f3cc))
- Rename tab by [@ymgyt](https://github.com/ymgyt) ([be4add1e](https://github.com/ymgyt/syndicationd/commit/be4add1e261c505d87b174795274236fd8ce46e7))

### Testing

- Impl device flow test case by [@ymgyt](https://github.com/ymgyt) ([93572902](https://github.com/ymgyt/syndicationd/commit/9357290265a4fbf8d78721e4f9f1904b1cf5b12a))
- Add auth flow case by [@ymgyt](https://github.com/ymgyt) ([6d2b1905](https://github.com/ymgyt/syndicationd/commit/6d2b1905d9b06bd9ed670f210cd590f89405c37c))
- Run kvsd in test by [@ymgyt](https://github.com/ymgyt) ([923e65a1](https://github.com/ymgyt/syndicationd/commit/923e65a131bed1a0a10d073b0eb9d5091cc184fe))
- Run integration test by [@ymgyt](https://github.com/ymgyt) ([20c0bc2d](https://github.com/ymgyt/syndicationd/commit/20c0bc2d31a938d3103fafedba5a10b4a9bba9ae))
- Fix tls conf path by [@ymgyt](https://github.com/ymgyt) ([e3d764a4](https://github.com/ymgyt/syndicationd/commit/e3d764a453b527a98b1eaf268ead67469c0e192d))

<!-- generated by git-cliff -->
