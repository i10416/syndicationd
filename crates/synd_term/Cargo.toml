[package]
authors.workspace     = true
categories.workspace  = true
description.workspace = true
edition.workspace     = true
keywords.workspace    = true
license.workspace     = true
readme.workspace      = true
repository.workspace  = true

name    = "synd-term"
version = "0.1.1"

[[bin]]
name = "synd"
path = "src/main.rs"

[dependencies]
synd-auth = { path = "../synd_auth", version = "0.1.2" }
synd-feed = { path = "../synd_feed", version = "0.1.1" }
synd-o11y = { path = "../synd_o11y", version = "0.1.1" }

anyhow         = { workspace = true }
chrono         = { workspace = true }
clap           = { workspace = true, features = ["derive", "string", "color", "suggestions", "wrap_help"] }
crossterm      = { version = "0.27.0", features = ["event-stream"] }
directories    = "5.0.1"
edit           = "0.1.5"
futures-util   = "0.3.30"
graphql_client = { workspace = true }
html2text      = { version = "0.12" }
open           = "5.0.1"
# ratatui              = { git = "https://github.com/ratatui-org/ratatui.git", branch = "main" }
ratatui              = { version = "0.26.0" }
reqwest              = { workspace = true }
serde                = { workspace = true, features = ["derive"] }
serde_json           = "1.0.111"
tokio                = { workspace = true, features = ["macros", "rt-multi-thread", "sync", "time"] }
tracing              = { workspace = true }
tracing-appender     = "0.2.3"
tracing-subscriber   = { workspace = true }
unicode-segmentation = "1.10.1"
url                  = { workspace = true }
# https://github.com/arkbig/throbber-widgets-tui/pull/5
# throbber-widgets-tui = "0.3.0"

[features]
# Integration test
integration = []

[dev-dependencies]
synd-api  = { path = "../synd_api" }
synd-test = { path = "../synd_test" }

axum-server  = { workspace = true }
kvsd         = { workspace = true }
serial_test  = { version = "3.0.0", default_features = false, features = ["async", "file_locks"] }
tempdir      = "0.3.7"
tokio-stream = "0.1.14"

[lints]
workspace = true
