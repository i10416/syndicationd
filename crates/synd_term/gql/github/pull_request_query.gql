query PullRequestQuery($repositoryOwner: String!, $repositoryName: String!, $pullRequestNumber: Int!) {
   repository(owner: $repositoryOwner, name: $repositoryName, followRenames: true) {
    repositoryTopics(first: 10) {
      nodes {
        topic {
          name
        }
      }
    }
    pullRequest(number: $pullRequestNumber) {
      totalCommentsCount
      state
      isDraft
      bodyText
      author {
        __typename
        login
      }
      labels(first: 10, orderBy: {field: NAME, direction: ASC}) {
        nodes {
          color
          name
        }
      }
      comments(first: 1, orderBy: {field: UPDATED_AT, direction: DESC}) {
        nodes {
          bodyText
          author {
            __typename
            login
          }
        }
      }
    }
  }
}
