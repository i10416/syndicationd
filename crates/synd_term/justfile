# List recipe
default:
    just --list

# Generate CHANGELOG for given version release 
changelog version:
    git cliff \
        --config ../../cliff.toml \
        --repository ../../ \
        --include-path crates/synd_term/** \
        --tag-pattern synd-term-v.* \
        --tag synd-term-v{{version}} \
        --output CHANGELOG.md

# Release crate
[no-cd]
release *flags:
    cargo release --package synd-term -v {{flags}}

# Reinstall synd-term
[no-cd]
reinstall:
    @nu etc/scripts/reinstall.nu
