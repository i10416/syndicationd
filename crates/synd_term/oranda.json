{
  "project": {
    "name": "synd"
  },
  "build": {
    "static_dir": "assets",
    "dist_dir": "website"
  },
  "styles": {
    "theme": "axo_dark",
    "favicon": "https://blog.ymgyt.io/favicon.ico"
  },
  "marketing": {
    "social": {
      "image": "https://blog.ymgyt.io/images/emoji/crab.png",
      "image_alt": "synd",
      "twitter_account": "@YAmaguchixt"
    }
  },
  "components": {
    "changelog": {
      "read_changelog_file": false
    },
    "artifacts": {
      "package_managers": {
        "additional": {
          "nix": "nix profile install github:ymgyt/syndicationd",
          "pacman": "pacman -S syndicationd",
          "cargo": "cargo install synd-term --locked",
          "docker": "docker run -it ghcr.io/ymgyt/synd-term"
        }
      }
    }
  }
}
