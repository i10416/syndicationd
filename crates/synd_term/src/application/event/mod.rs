use crate::command::Command;

mod key_handlers;
pub use key_handlers::{KeyHandler, KeyHandlers};

pub(crate) enum KeyEventResult {
    Consumed(Option<Command>),
    Ignored,
}
