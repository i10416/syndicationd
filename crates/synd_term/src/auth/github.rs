use std::{io::Write, time::Duration};

use http::StatusCode;
use reqwest::Client;
use tracing::debug;

use crate::{
    auth::device_flow::{
        DeviceAccessTokenErrorResponse, DeviceAccessTokenRequest, DeviceAccessTokenResponse,
        DeviceAuthorizationRequest, DeviceAuthorizationResponse,
    },
    config,
};

/// https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#device-flow
#[derive(Clone)]
pub struct DeviceFlow {
    client: Client,
    client_id: &'static str,
    endpoint: Option<&'static str>,
}

impl DeviceFlow {
    const DEVICE_AUTHORIZATION_ENDPOINT: &'static str = "https://github.com/login/device/code";
    const TOKEN_ENDPOINT: &'static str = "https://github.com/login/oauth/access_token";

    pub fn new() -> Self {
        let client = reqwest::ClientBuilder::new()
            .user_agent(config::USER_AGENT)
            .timeout(Duration::from_secs(5))
            .build()
            .unwrap();

        Self {
            client,
            client_id: config::github::CLIENT_ID,
            endpoint: None,
        }
    }

    pub fn with_endpoint(self, endpoint: &'static str) -> Self {
        Self {
            endpoint: Some(endpoint),
            ..self
        }
    }

    #[tracing::instrument(skip(self))]
    pub async fn device_authorize_request(&self) -> anyhow::Result<DeviceAuthorizationResponse> {
        tracing::debug!("Sending request");

        // https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps
        let scope = "user:email";

        let response = self
            .client
            .post(self.endpoint.unwrap_or(Self::DEVICE_AUTHORIZATION_ENDPOINT))
            .header(http::header::ACCEPT, "application/json")
            .form(&DeviceAuthorizationRequest {
                client_id: self.client_id.into(),
                scope: scope.into(),
            })
            .send()
            .await?
            .error_for_status()?
            .json::<DeviceAuthorizationResponse>()
            .await?;

        tracing::debug!("Got response");

        Ok(response)
    }

    pub async fn pool_device_access_token(
        &self,
        device_code: String,
        interval: Option<i64>,
    ) -> anyhow::Result<DeviceAccessTokenResponse> {
        // poll to check if user authorized the device
        macro_rules! continue_or_abort {
            ( $response_bytes:ident ) => {{
                let err_response = serde_json::from_slice::<DeviceAccessTokenErrorResponse>(&$response_bytes)?;
                if err_response.error.should_continue_to_poll() {
                    debug!(error_code=?err_response.error,interval, "Continue to poll");

                    let interval = interval.unwrap_or(5);

                    tokio::time::sleep(Duration::from_secs(interval as u64)).await;
                } else {
                    anyhow::bail!(
                        "Failed to authenticate. authorization server respond with {err_response:?}"
                    )
                }
            }};
        }

        let response = loop {
            let response = self
                .client
                .post(Self::TOKEN_ENDPOINT)
                .header(http::header::ACCEPT, "application/json")
                .form(&DeviceAccessTokenRequest::new(&device_code, self.client_id))
                .send()
                .await?;

            debug!("{:?}", response.status());

            match response.status() {
                StatusCode::OK => {
                    let full = response.bytes().await?;
                    match serde_json::from_slice::<DeviceAccessTokenResponse>(&full) {
                        Ok(response) => break response,
                        Err(_) => continue_or_abort!(full),
                    }
                }
                StatusCode::BAD_REQUEST => {
                    let full = response.bytes().await?;
                    continue_or_abort!(full)
                }
                other => {
                    let error_msg = response.text().await.unwrap_or_default();
                    anyhow::bail!("Failed to authenticate. authorization server respond with {other} {error_msg}")
                }
            }
        };

        Ok(response)
    }

    #[tracing::instrument(skip_all)]
    pub async fn device_flow<W: Write>(
        self,
        writer: W,
    ) -> anyhow::Result<DeviceAccessTokenResponse> {
        let DeviceAuthorizationResponse {
            device_code,
            user_code,
            verification_uri,
            interval,
            ..
        } = self.device_authorize_request().await?;

        let mut writer = writer;
        writeln!(&mut writer, "Open `{verification_uri}` on your browser")?;
        writeln!(&mut writer, "Enter CODE: `{user_code}`")?;

        // attempt to open input screen in the browser
        open::that(verification_uri.to_string()).ok();

        self.pool_device_access_token(device_code, interval).await
    }
}
