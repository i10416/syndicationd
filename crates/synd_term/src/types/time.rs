pub type Time = synd_feed::types::Time;

pub trait TimeExt {
    fn local_ymd(&self) -> String;
}

impl TimeExt for Time {
    fn local_ymd(&self) -> String {
        self.naive_local().format("%Y/%m/%d").to_string()
    }
}
