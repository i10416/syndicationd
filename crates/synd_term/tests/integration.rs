#[cfg(feature = "integration")]
mod test {
    mod helper;

    use std::time::Duration;

    use crossterm::event::{Event, KeyCode, KeyEvent};
    use ratatui::{
        prelude::Buffer,
        style::{Modifier, Style},
    };
    use serial_test::file_serial;

    use synd_authn::device_flow::github::DeviceFlow;

    use synd_term::{
        application::{Application, Config},
        client::Client,
        ui::theme::Theme,
    };
    use tokio::net::TcpListener;
    use tokio_stream::wrappers::UnboundedReceiverStream;
    use tracing_subscriber::EnvFilter;

    #[tokio::test(flavor = "multi_thread")]
    #[file_serial(a)]
    async fn hello_world() -> anyhow::Result<()> {
        // TODO: wrap once
        tracing_subscriber::fmt()
            .with_env_filter(EnvFilter::from_default_env())
            .with_line_number(true)
            .with_file(true)
            .with_target(false)
            .init();

        tracing::info!("TEST hello_world run");

        let mock_port = 6000;
        let api_port = 6001;
        let oauth_addr = ("127.0.0.1", mock_port);
        let oauth_listener = TcpListener::bind(oauth_addr).await?;
        tokio::spawn(synd_test::mock::serve(oauth_listener));
        helper::serve_api(mock_port, api_port).await?;

        let endpoint = format!("http://localhost:{api_port}/graphql")
            .parse()
            .unwrap();
        let terminal = helper::new_test_terminal();
        let client = Client::new(endpoint).unwrap();
        let config = Config {
            idle_timer_interval: Duration::from_millis(1000),
            throbber_timer_interval: Duration::from_secs(3600), // disable throbber
            github_device_flow: DeviceFlow::new("dummy")
                .with_device_authorization_endpoint(
                    "http://localhost:6000/case1/github/login/device/code",
                )
                .with_token_endpoint("http://localhost:6000/case1/github/login/oauth/access_token"),
        };
        // or mpsc and tokio_stream ReceiverStream
        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
        let mut event_stream = UnboundedReceiverStream::new(rx);
        let theme = Theme::new();
        let bg = theme.background.bg.unwrap_or_default();

        let mut application = Application::with(terminal, client, config).with_theme(theme.clone());
        application.event_loop_until_idle(&mut event_stream).await;

        // login
        #[rustfmt::skip]
        let mut expected = Buffer::with_lines(vec![
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                      Login                                     ",
            "                        ────────────────────────────────                        ",
            "                        >> with GitHub                                          ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
        ]);
        for y in 0..expected.area.height {
            for x in 0..expected.area.width {
                expected.get_mut(x, y).set_bg(bg);
            }
        }
        // title
        for x in 38..43 {
            expected
                .get_mut(x, 5)
                .set_style(Style::new().add_modifier(Modifier::BOLD));
        }
        // auth provider
        for x in 24..56 {
            expected
                .get_mut(x, 7)
                .set_style(Style::new().add_modifier(Modifier::BOLD));
        }

        application.assert_buffer(&expected);

        tracing::info!("Login assertion OK");

        // push enter => start auth flow
        let event = Event::Key(KeyEvent::from(KeyCode::Enter));
        tx.send(Ok(event)).unwrap();
        application.event_loop_until_idle(&mut event_stream).await;

        // assert login prompt
        #[rustfmt::skip]
        let mut expected = Buffer::with_lines(vec![
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                      Login                                     ",
            "                        ────────────────────────────────                        ",
            "                        Open the following URL and Enter                        ",
            "                                                                                ",
            "                        URL:  https://syndicationd.ymgyt                        ",
            "                        Code: UC123456                                          ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
        ]);
        for y in 0..expected.area.height {
            for x in 0..expected.area.width {
                expected.get_mut(x, y).set_bg(bg);
            }
        }
        // title
        for x in 38..43 {
            expected
                .get_mut(x, 5)
                .set_style(Style::new().add_modifier(Modifier::BOLD));
        }
        // Bold url
        for x in 30..56 {
            expected
                .get_mut(x, 9)
                .set_style(Style::new().add_modifier(Modifier::BOLD));
        }
        // Bold code
        for x in 30..38 {
            expected
                .get_mut(x, 10)
                .set_style(Style::new().add_modifier(Modifier::BOLD));
        }

        application.assert_buffer(&expected);

        tracing::info!("Login prompt assertion OK");

        // for quit event loop after polling job complete
        application.reset_idle_timer();
        // polling device access token complete
        application.event_loop_until_idle(&mut event_stream).await;

        #[rustfmt::skip]
        let mut expected = Buffer::with_lines(vec![
            "    Syndicationd                                         Feeds    Subscription  ",
            "                                                                                ",
            "    Published   Title                                      Feed                 ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━Summary━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "                                                                                ",
            "              q  Quit  Tab  Next Tab  j/k  Up/Down  Ent  Open Entry             "
        ]);
        for y in 0..expected.area.height {
            for x in 0..expected.area.width {
                expected.get_mut(x, y).set_bg(bg);
            }
        }
        // tab line
        {
            // syndicationd
            for x in 4..16 {
                expected.get_mut(x, 0).set_style(theme.application_title);
            }
            // feeds tab left padding
            for x in 53..57 {
                expected.get_mut(x, 0).set_style(theme.tabs);
            }
            // selected feeds tab
            for x in 57..62 {
                expected.get_mut(x, 0).set_style(theme.tabs_selected);
            }
            // left padding and subscription
            for x in 62..78 {
                expected.get_mut(x, 0).set_style(theme.tabs);
            }
        }
        // table header
        {
            for x in 1..expected.area.width - 1 {
                expected.get_mut(x, 2).set_style(theme.entries.header);
            }
        }
        // prompt
        {
            // q
            for x in 13..16 {
                expected.get_mut(x, 19).set_style(theme.prompt.key);
            }
            // Quit
            for x in 16..22 {
                expected.get_mut(x, 19).set_style(theme.prompt.key_desc);
            }
            // Tab
            for x in 22..27 {
                expected.get_mut(x, 19).set_style(theme.prompt.key);
            }
            // Next Tab
            for x in 27..37 {
                expected.get_mut(x, 19).set_style(theme.prompt.key_desc);
            }
            // j/k
            for x in 37..42 {
                expected.get_mut(x, 19).set_style(theme.prompt.key);
            }
            // Up/Down
            for x in 42..51 {
                expected.get_mut(x, 19).set_style(theme.prompt.key_desc);
            }
            // Ent
            for x in 51..56 {
                expected.get_mut(x, 19).set_style(theme.prompt.key);
            }
            // Open entry
            for x in 56..68 {
                expected.get_mut(x, 19).set_style(theme.prompt.key_desc);
            }
        }

        application.assert_buffer(&expected);

        Ok(())
    }
}
