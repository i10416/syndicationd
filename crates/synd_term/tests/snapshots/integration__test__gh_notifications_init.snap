---
source: crates/synd_term/tests/integration.rs
description: github notifications initial
expression: application.buffer()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 120, height: 30 },
    content: [
        "  Syndicationd                                                                         󰊤 GitHub    󱉯 Entries    󰑫 Feeds ",
        "  󰈶 Filter  Unread  Participating                                                                                      ",
        "   Search                                                                                                              ",
        "                                                                                                                        ",
        "  Updated   Title 1/2                                                                                 Reposi  Reason    ",
        "  1h ago    title AA1                                                                               repo-a  mentioned▐",
        "   2h ago    title AA2                                                                               repo-a  mentioned ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "                                                                                                                        ",
        "────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────",
        "   Subject    sakura / repo-a #1                                                                                       ",
        "  󰯂 Title      title AA1                                                                                                ",
        "   UpdatedAt  2024-07-05 07:00 (+00:00)                                                                                ",
        "  󱍵 Labels     A-label C-enhancement                                                                                    ",
        "  󱙓 Issue      @userX                                                                                                   ",
        "  Hello World                                                                                                           ",
        "   Comment    @userY                                                                                                   ",
        "  LastComment...                                                                                                        ",
        "                          Tab:󰹳  j/k:󰹹  gg:󱞧  ge:󱞥  c:  /:  f:󰈶  Ent:󰏌  d:  u:  r:󰑓  q:                            ",
    ],
    styles: [
        x: 0, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 14, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 87, y: 0, fg: Rgb(255, 160, 122), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 95, y: 0, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 10, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 20, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 33, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 35, y: 1, fg: Rgb(247, 76, 0), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 36, y: 1, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 2, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 10, y: 2, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 0, y: 4, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD | UNDERLINED,
        x: 0, y: 5, fg: Rgb(255, 160, 122), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 119, y: 5, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 0, y: 6, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 12, y: 6, fg: Green, bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 13, y: 6, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 21, fg: Green, bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 3, y: 21, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 11, y: 21, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 22, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 9, y: 22, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 23, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 13, y: 23, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 24, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD,
        x: 10, y: 24, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 15, y: 24, fg: Rgb(43, 41, 45), bg: Rgb(251, 202, 4), underline: Reset, modifier: NONE,
        x: 22, y: 24, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 23, y: 24, fg: Rgb(43, 41, 45), bg: Rgb(197, 222, 245), underline: Reset, modifier: NONE,
        x: 36, y: 24, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD | UNDERLINED,
        x: 9, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 21, y: 25, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 2, y: 27, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: BOLD | UNDERLINED,
        x: 11, y: 27, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 14, y: 27, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: DIM,
        x: 21, y: 27, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 26, y: 29, fg: Rgb(111, 93, 99), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
        x: 94, y: 29, fg: Rgb(254, 205, 178), bg: Rgb(43, 41, 45), underline: Reset, modifier: NONE,
    ]
}
